tests/unit/store/async-leak-test.js

```diff
@@ -0,0 +1,359 @@
+            }, 50); // intentionally longer than the 10ms polling interval of `wait()`
+          });
+        },
+      })
+    );
+
+    let request = store.findRecord('person', '1');
+    let waiter = store.__asyncWaiter;
+
+    assert.equal(
+      waiter(),
+      true,
+      'We return true when no requests have been initiated yet (pending queue flush is async)'
+    );
+
+    await stepPromise;
+
+    assert.equal(waiter(), false, 'We return false to keep waiting while requests are pending');
+
+    await request;
+
+    assert.equal(waiter(), true, 'We return true to end waiting when no requests are pending');
+  });
+
+  test('waiter can be turned off', async function(assert) {
+    let stepResolve;
+    let stepPromise = new Promise(resolveStep => {
+      stepResolve = resolveStep;
+    });
+    this.owner.register(
+      'adapter:application',
+      JSONAPIAdapter.extend({
+        findRecord() {
+          return new Promise(resolve => {
+            stepResolve();
+
+            setTimeout(() => {
+              resolve({ data: { type: 'person', id: '1' } });
+            }, 50); // intentionally longer than the 10ms polling interval of `wait()`
+          });
+        },
+      })
+    );
+
+    // turn off the waiter
+    store._shouldTrackAsyncRequests = false;
+
+    let request = store.findRecord('person', '1');
+    let waiter = store.__asyncWaiter;
+
+    assert.equal(
+      waiter(),
+      true,
+      'We return true when no requests have been initiated yet (pending queue flush is async)'
+    );
+
+    await stepPromise;
+
+    assert.equal(
+      store._trackedAsyncRequests.length,
+      1,
+      'We return true even though a request is pending'
+    );
+    assert.equal(waiter(), true, 'We return true even though a request is pending');
+
+    await request;
+
+    assert.equal(waiter(), true, 'We return true to end waiting when no requests are pending');
+  });
+
+  test('waiter works even when the adapter rejects', async function(assert) {
+    assert.expect(4);
+    let stepResolve;
+    let stepPromise = new Promise(resolveStep => {
+      stepResolve = resolveStep;
+    });
+    this.owner.register(
+      'adapter:application',
+      JSONAPIAdapter.extend({
+        findRecord() {
+          return new Promise((resolve, reject) => {
+            stepResolve();
+
+            setTimeout(() => {
+              reject({ errors: [] });
+            }, 50); // intentionally longer than the 10ms polling interval of `wait()`
+          });
+        },
+      })
+    );
+
+    let request = store.findRecord('person', '1');
+    let waiter = store.__asyncWaiter;
+
+    assert.equal(
+      waiter(),
+      true,
+      'We return true when no requests have been initiated yet (pending queue flush is async)'
+    );
+
+    await stepPromise;
+
+    assert.equal(waiter(), false, 'We return false to keep waiting while requests are pending');
+
+    await request.catch(e => {
```

Comment:
`assert.rejects` if present in this version of QUnit, could be used to provide better test failure messages.

more context: https://github.com/qunitjs/qunit/pull/1238/files#diff-78b865bceaf70054fdd9917848e7ddd8R43